wgpu-native/src/instance.rs

```diff
@@ -301,11 +302,34 @@ pub extern "C" fn wgpu_create_surface_from_windows_hwnd(
     ))
 }
 
-pub fn request_adapter(
+pub type RequestAdapterCallback =
+    extern "C" fn(adapter: *const AdapterId, userdata: *mut c_void);
+
+pub fn request_adapter_async(
     global: &Global,
     desc: &RequestAdapterOptions,
     input_ids: &[Input<AdapterId>],
-) -> Option<AdapterId> {
+    callback: RequestAdapterCallback,
+    userdata: *mut c_void,
+) {
+    let picked = pick_adapter(global, desc, input_ids);
+    #[cfg(not(feature = "local"))]
+    let (_, adapter) = picked;
+    #[cfg(feature = "local")]
+    let (adapter, _) = picked;
+    callback(adapter.as_ref().map_or(&AdapterId::ERROR, |x| x as *const _), userdata);
+}
+
+#[cfg(not(feature = "local"))]
+type PickedAdapter = (Option<std::marker::PhantomData<AdapterId>>, Option<AdapterId>);
+#[cfg(feature = "local")]
+type PickedAdapter = (Option<AdapterId>, Option<std::marker::PhantomData<AdapterId>>);
+
+fn pick_adapter(
+    global: &Global,
+    desc: &RequestAdapterOptions,
+    input_ids: &[Input<AdapterId>],
+) -> PickedAdapter {
```

Comment:
Can we just return `Option<AdapterId>`?

Reply:
@kvark that's a good idea, initially I was confused about the purpose of both IDs here. Updated the commit